segmentation.py

```diff
@@ -147,69 +157,76 @@ def add_adjacent_connector_column(component, next_component, schematic):
-                                               Dict[int, Dict[int, set]], Set[int]]:
+def find_dividers(matrix: List[Path]) -> Tuple[pd.DataFrame, Set[int]]:
     max_bin = 1
-    leaving = nested_dict(2, set)  # containing the set of participating Paths on that link column
-    entering = nested_dict(2, set)  # list of indices of new components
-    dividers = {1}  # all start positions of components, start with st
-    copy_arrivals = set()  # track self loops just in case component gets cut in half
-    uniq_links = set()
+
+    self_loops = []  # track self loops just in case component gets cut in half
+    connection_dfs = []  # pandas dataframe with columns (from, to, path [name])
+
     for i, path in enumerate(matrix):
-        print(f"Segmenting path {path.name} with number of bins {len(path.bins)} and total progress:", '{0:.1%}'.format(i / len(matrix)))
-        max_bin = max(max_bin, max(path._bin_set))
-        for link in path.links:  # Links are generated by odgi based
-            upstream, downstream = link.upstream, link.downstream
-            # Links to 0 Bin indicate the beginning or end of a path.  0 Bin has no sequence
-            if 0 in [upstream, downstream]:
-                continue  # ignore links to the telomere.
-            if upstream == downstream:
-                copy_arrivals.add(upstream)
-                continue  # we don't want these to become dividers
-            # Is the gap range anywhere else in this individual?
-            # What if downstream < upstream?
-            divider_verified = downstream < upstream
-            if not divider_verified:
-                missing_range = list(range(upstream + 1, downstream))
-                for i in missing_range:
-                    if i in path:
-                        divider_verified = True
-                        uniq_links.add((upstream, downstream))
-                        break  # stop as soon as we have confirmation
-            if divider_verified:
-                # if (upstream + 1) in leaving.keys() :
-                #     print(f"Found inherited rearrangement {upstream+1}")
-                leaving[upstream][downstream].add(path.name)  # the first position of the new component
-                entering[downstream][upstream].add(path.name)
-                dividers.add(upstream + 1)
-                dividers.add(downstream)
-                # TODO: insert prevarications about exact position
-                # Divider should be somewhere in here
-                # Tolerable range?
-                # Stack up others using the same LinkColumn
+        links = path.links
+        bin_ids = np.array([b.bin_id for b in path.bins])
+        bin_ids.sort()
+
+        if bin_ids.size > 0:
+            max_bin = max(max_bin, int(bin_ids[-1]))
+
+        # we don't want these to become dividers
+        boundary_mask = utils.path_boundaries(links)
+        self_loops_mask = utils.self_loops(links)
+
+        if np.any(self_loops_mask):
```

Comment:
Based on my recollection, self_loops turned out to not exist in odgi output.  I was testing for them, but they don't happen so you could save some time and complexity by removing the concept.  (That can be in a later PR)

Reply:
Thanks, good to know.